File: 5. Longest Palindromic Substring/5.cpp
#include <deque>
class Solution {
public:
    string longestPalindrome(string s) {
        deque<char> deq;
        int len = s.length();
        string output;
        int currentLen = 0, maxLen = 0, lptr = 0, rptr = 0;
        for (int i=0; i<len; i++){
            for (int y=0; y<2; y++){
                deq.push_back(s[i]);
                currentLen = 1;
                lptr = i;
                rptr = i;
                if (y == 1){
                    if ((i+1) <= (len-1)){
                        if (s[i] == s[i+1]){
                            deq.push_back(s[i]);
                            currentLen = 2;
                            rptr++;
                        }
                    }
                    else
                        break;
                }
                while ((--lptr >= 0) && (++rptr <= (len-1)))
                {
                    if (s[lptr] == s[rptr]){
                        currentLen += 2;
                        deq.push_front(s[lptr]);
                        deq.push_back(s[rptr]);
                    }
                    else 
                        break;
                }
                if (currentLen > maxLen){
                    maxLen = currentLen;
                    output.clear();
                    while (!deq.empty()){
                        output += deq.front();
                        deq.pop_front();
                    }

                }
                deq.clear();
            }
        }
        return output;
    }
};

File: 5. Longest Palindromic Substring/5.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        answer_lptr = 0
        answer_rptr = 0
        max_length = 0
        for i, c in enumerate(s):
            lptr = rptr = i
            while rptr < len(s) - 1 and s[rptr+1] == c:
                rptr += 1
            while lptr > 0 and rptr < len(s) - 1:
                if s[lptr-1]  == s[rptr+1]:
                    lptr -= 1
                    rptr += 1
                else:
                    break
            current_length = rptr - lptr + 1
            if current_length > max_length:
                max_length = current_length
                answer_lptr = lptr
                answer_rptr = rptr

        return s[answer_lptr:answer_rptr+1]
    
s = 'bb'
solve = Solution().longestPalindrome(s)
print(solve)

        